File: packages/zodiac-js/src/calculator/indian-calculator.ts
import { INDIAN_RASHIS } from '../indian/rashis';
import { validateDate } from '../validation/validator';
import { getLocaleText, getSupportedLocales } from '../localization/locale-manager';
import { getIndianIcon } from '../icons/icon-provider';
import type { IndianResult } from '../models/indian-rashi';

export interface IndianOptions {
  locale?: string;
  iconMode?: 'none' | 'builtin' | 'custom';
}

/**
 * Determines the Indian Zodiac sign (Rashi) for a given month and day.
 */
export function getIndianZodiac(month: number, day: number, options?: IndianOptions): IndianResult {
  validateDate(month, day);

  let foundRashi = INDIAN_RASHIS[0]; // fallback, shouldn't occur

  for (const rashi of INDIAN_RASHIS) {
    const { start, end } = rashi.dateRange;

    // e.g. Start: Apr 13, End: May 14
    if (
      (month === start.month && day >= start.day) ||
      (month === end.month && day <= end.day)
    ) {
      foundRashi = rashi;
      break;
    }
  }

  const result: Partial<IndianResult> = { rashi: foundRashi };

  if (options?.locale) {
    const locale = options.locale;
    (result as any).locale = getSupportedLocales().includes(locale) ? locale : 'en';
    (result as any).localizedName = getLocaleText(locale, `indian.${foundRashi.id}.name`) ?? foundRashi.name;
  }

  const iconMode = options?.iconMode ?? 'builtin';
  if (iconMode !== 'none') {
    result.icon = getIndianIcon(foundRashi.id, foundRashi.westernEquivalent, iconMode);
  }

  return result as IndianResult;
}

File: packages/zodiac-dart/lib/src/calculator/indian_calculator.dart
import '../models/indian_rashi.dart';
import '../indian/rashis.dart';
import '../validation/validator.dart';
import '../localization/locale_manager.dart';
import '../icons/icon_provider.dart';
import '../models/zodiac_icon.dart';

class IndianOptions {
  final String? locale;
  final String iconMode;

  const IndianOptions({
    this.locale,
    this.iconMode = 'builtin',
  });
}

/// Determines the Indian Zodiac sign (Rashi) for a given month and day.
IndianResult getIndianZodiac(int month, int day, {IndianOptions? options}) {
  validateDate(month, day);

  var foundRashi = indianRashis[0];

  for (var rashi in indianRashis) {
    final startMonth = rashi.dateRange.startMonth;
    final startDay = rashi.dateRange.startDay;
    final endMonth = rashi.dateRange.endMonth;
    final endDay = rashi.dateRange.endDay;

    if ((month == startMonth && day >= startDay) ||
        (month == endMonth && day <= endDay)) {
      foundRashi = rashi;
      break;
    }
  }

  String? resolvedLocale;
  String? localizedName;

  if (options?.locale != null) {
    final locale = options!.locale!;
    resolvedLocale = getSupportedLocales().contains(locale) ? locale : 'en';

    localizedName =
        getLocaleText(locale, 'indian.${foundRashi.id}.name') as String? ??
            foundRashi.name;
  }

  ZodiacIcon? icon;
  final iconMode = options?.iconMode ?? 'builtin';
  if (iconMode != 'none') {
    icon = getIndianIcon(foundRashi.id, foundRashi.westernEquivalent, iconMode);
  }

  return IndianResult(
    rashi: foundRashi,
    locale: resolvedLocale,
    localizedName: localizedName,
    icon: icon,
  );
}
